applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_analysis.py

```diff
@@ -0,0 +1,274 @@
+        while self.time < self.end_time:
+            self.InitializeTimeStep()
+            self.SolveTimeStep()
+            self.FinalizeTimeStep()
+
+    #### Public functions defining the Interface to the CoSimulationApplication ####
+    def Initialize(self, external_model_part=None):
+        self.ImportAndCreateSolver(external_model_part)
+        self.InitializeIO()
+        self.ExecuteInitialize()
+        self.ExecuteBeforeSolutionLoop()
+
+    def InitializeTimeStep(self):
+        self.ExecuteInitializeSolutionStep()
+
+    def SolveTimeStep(self):
+        self.SolveSolutionStep();
+
+    def FinalizeTimeStep(self):
+        self.ExecuteFinalizeSolutionStep()
+
+    def Finalize(self):
+        self.ExecuteFinalize()
+
+
+    ###########################################################################
+    def ImportAndCreateSolver(self, external_model_part=None):
+        """ Importing the Solver and the ModelPart """
+        if external_model_part != None:
+            # This is a temporary solution until the importing of the ModelPart
+            # is removed from the solver (needed e.g. for Optimization)
+            if (type(external_model_part) != KratosMultiphysics.ModelPart):
+                raise Exception("Input is expected to be provided as a Kratos ModelPart object")
+            using_external_model_part = True
+        else:
+            using_external_model_part = False
+
+        ## Get echo level and parallel type
+        self.echo_level = self.ProjectParameters["problem_data"]["echo_level"].GetInt()
+        self.parallel_type = self.ProjectParameters["problem_data"]["parallel_type"].GetString()
+
+        # To avoid many prints # TODO leave this?
+        if (self.echo_level == 0):
+            KratosMultiphysics.Logger.GetDefaultOutput().SetSeverity(KratosMultiphysics.Logger.Severity.WARNING)
+
+        ## Import parallel modules if needed
+        if (self.parallel_type == "MPI"):
+            import KratosMultiphysics.mpi as KratosMPI
+            import KratosMultiphysics.MetisApplication as MetisApplication
+            import KratosMultiphysics.TrilinosApplication as TrilinosApplication
+
+        ## Structure model part definition
+        if using_external_model_part:
+            self.main_model_part = external_model_part
+        else:
+            main_model_part_name = self.ProjectParameters["problem_data"]["model_part_name"].GetString()
+            self.main_model_part = KratosMultiphysics.ModelPart(main_model_part_name)
+            self.main_model_part.ProcessInfo.SetValue(KratosMultiphysics.DOMAIN_SIZE,
+                                                      self.ProjectParameters["problem_data"]["domain_size"].GetInt())
+
+        ## Solver construction
+        import python_solvers_wrapper_structural
+        self.solver = python_solvers_wrapper_structural.CreateSolver(self.main_model_part, self.ProjectParameters)
+
+        if not using_external_model_part:
```

Comment:
i believe you might want to AddVariable and AddDofs even for an external modelpart. At least in a future...

Reply:
I discussed this extensively with @armingeiser but we haven't though abt what you are proposing.
It sounds good, I will discuss it with him tmr, thx for the suggestion though